lib/dal/src/migrations/U2424__optimize_component_creation.sql

```diff
@@ -0,0 +1,365 @@
+      this_visibility,
+      'null'::jsonb,
+      this_unset_func_id
+    );
+
+    -- External Providers
+    FOR this_external_provider IN
+        SELECT *
+        FROM external_providers_v1(this_tenancy, this_visibility)
+        WHERE schema_variant_id = this_schema_variant_id
+    LOOP
+        this_attribute_context := attribute_context_build_from_parts_v1(
+            ident_nil_v1(), -- Prop ID
+            ident_nil_v1(), -- Internal Provider ID
+            this_external_provider.id, -- External Provider ID
+            -- We won't find a component-specific prototype, since the component
+            -- didn't exist before calling this function, but we'll want the
+            -- component ID set when we go to create the Attribute Value.
+            this_new_row.id -- Component ID
+        );
+
+        SELECT *
+        INTO STRICT this_attribute_prototype
+        FROM attribute_prototypes_v1(this_tenancy, this_visibility) AS ap
+        WHERE in_attribute_context_v1(this_attribute_context, ap);
+
+        SELECT av.object
+        INTO this_new_attribute_value
+        FROM attribute_value_create_v1(
+            this_tenancy,
+            this_visibility,
+            this_attribute_context,
+            this_unset_func_binding_id,
+            this_unset_func_binding_return_value_id,
+            NULL
+        ) AS av;
+
+        PERFORM set_belongs_to_v1(
+            'attribute_value_belongs_to_attribute_prototype',
+            this_tenancy,
+            this_visibility,
+            this_new_attribute_value ->> 'id',
+            this_attribute_prototype.id
+        );
+    END LOOP;
+
+    -- Explicit Internal Providers
+    FOR this_internal_provider IN
+        SELECT *
+        FROM internal_providers_v1(this_tenancy, this_visibility)
+        WHERE schema_variant_id = this_schema_variant_id
+    LOOP
+        this_attribute_context := attribute_context_build_from_parts_v1(
+            ident_nil_v1(), -- Prop ID
+            this_internal_provider.id, -- Internal Provider ID
+            ident_nil_v1(), -- External Provider ID
+            this_new_row.id -- Component ID
+        );
+
+        SELECT *
+        INTO STRICT this_attribute_prototype
+        FROM attribute_prototypes_v1(this_tenancy, this_visibility) AS ap
+        WHERE in_attribute_context_v1(this_attribute_context, ap);
+
+        SELECT av.object
+        INTO this_new_attribute_value
+        FROM attribute_value_create_v1(
+            this_tenancy,
+            this_visibility,
+            this_attribute_context,
+            this_unset_func_binding_id,
+            this_unset_func_binding_return_value_id,
+            NULL
+        ) AS av;
+
+        PERFORM set_belongs_to_v1(
+            'attribute_value_belongs_to_attribute_prototype',
+            this_tenancy,
+            this_visibility,
+            this_new_attribute_value ->> 'id',
+            this_attribute_prototype.id
+        );
+    END LOOP;
+
+    SELECT array_agg(ip.*)
```

Comment:
The significant changes to this function happened from this line up to the foreach (the other changes were to use the values we stored in the arrays instead of selecting from the database)